mempool/src/pool.rs

```diff
@@ -0,0 +1,903 @@
+// object
+fn compare_fees(this: &TxMempoolEntry, other: &TxMempoolEntry) -> Ordering {
+    // We want the comparison to be such that a collection is sorted if its elements are ordered in
+    // descending order
+    other.get_fee().cmp(&this.get_fee())
+}
+
+impl TxMempoolEntry {
+    fn new(tx: Transaction, parents: BTreeSet<Rc<TxMempoolEntry>>, is_replaceable: bool) -> Self {
+        Self {
+            tx,
+            parents,
+            is_replaceable,
+        }
+    }
+    //TODO this should really be sum of inputs minus sum of outputs
+    //But we don't yet have a way of summing of the inputs
+    //TODO handle the error bettter
+    fn get_fee(&self) -> Amount {
+        self.tx
+            .get_outputs()
+            .iter()
+            .map(|output| output.get_value())
+            .sum::<Option<_>>()
+            .expect("fee overflow")
+    }
+}
+
+impl PartialOrd for TxMempoolEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(compare_fees(self, other))
+    }
+}
+
+impl Ord for TxMempoolEntry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        compare_fees(self, other)
+    }
+}
+
+pub(crate) struct MempoolImpl<C: ChainStateView> {
+    store: MempoolStore,
+    chain_state: C,
+}
+
+struct MempoolStore {
+    // TODO some cloning can be avoided by having spender_txs be a map between references
+    txs_by_id: HashMap<H256, Rc<TxMempoolEntry>>,
+    txs_by_fee: BTreeSet<Rc<TxMempoolEntry>>,
+    spender_txs: HashMap<OutPoint, Rc<TxMempoolEntry>>,
+}
+
+impl MempoolStore {
+    fn new() -> Self {
+        Self {
+            txs_by_fee: BTreeSet::new(),
+            txs_by_id: HashMap::new(),
+            spender_txs: HashMap::new(),
+        }
+    }
+
+    // Checks whether the outpoint belongs to one of the txs in the mempool
+    fn contains_outpoint(&self, outpoint: &OutPoint) -> bool {
+        matches!(self.txs_by_id.get(&outpoint.get_tx_id().get()),
+            Some(entry) if entry.tx.get_inputs().len() > outpoint.get_output_index() as usize)
+    }
+
+    fn create_entry(&self, tx: Transaction) -> TxMempoolEntry {
+        let parents = tx
+            .get_inputs()
+            .iter()
+            .filter_map(|input| self.txs_by_id.get(&input.get_outpoint().get_tx_id().get()))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+
+        // We implement what is known in BIP125 as "inherited signalling":
+        // A Transaction is replaceable if either (1) it signals replaceability explicitly (via its flags), OR (2) one of its in-mempool ancestors explicit signals replaceability
+
+        // At the level of TxMempoolEntry, signalling replaceability is always explicit, via the
+        // `is_replaceable` field. By construction, to determine the value of this field it is
+        // enoufh to scan direct parents, not all ancestors.
+        let is_replaceable =
+            tx.is_replaceable() || parents.iter().any(|parent| parent.is_replaceable);
```

Comment:
Now I realised there might be another potential problem with this approach. Say we have a graph like this:
```
TX1 [replaceable] <- TX2
```
That is, `TX1` is marked replaceable and there is `TX2` which spends `TX1` and is not explicitly marked replaceable. `TX2` is implicitly replaceable because it spends a replaceable transaction.

Now suppose a new block comes in and the block includes `TX1` but `TX2` stays in the mempool. The state of the mempool is now:
```
TX2
```

Should `TX2` still be replaceable? My instinct is that it should no longer be replaceable, otherwise all the descendant transactions of `TX1` may end up being replaceable, potentially forever, but I am not quite sure here.

It may be helpful to view the `is_replaceable` flag in the pool entry as cache. At any point, you can remove these flags and recalculate them from the scratch from the transaction graph. Also at any point `entry.is_replaceable` should give the same result as recalculating it from the transaction graph afresh. That means the entry flag has to be invalidated and eventually recalculated whenever there is a change that may affect the result. Moving transaction from the mempool to a block is quite plausibly an example of such a change.

Thoughts / comments?

Reply:
>otherwise all the descendant transactions of TX1 may end up being replaceable, potentially forever, but I am not quite sure here.

Is it possible to mark, e.g., `TX3`, which depends on `TX2`, irreplaceable and thus not worry about forever replaceable descendants? Like the default (if you don't specify anything) is that it's inherited from parents and that's overridden if you specify that this TX must be/must not be replaceable. I haven't read the BIP so I don't know if this suggestion even makes sense.

Reply:
My thoughts before I go diving into the BIP.
1. The main question here is not about our implementation per se but about our understanding of BIP125: i.e., is replaceability a "static" or "dynamic" property? 
2. Let's avoid all forms of caching if possible. If it turns out replaceability is a dynamic property then lets just calculate it whenever necessary. As the saying goes, there are two hard problems in computer science, and caching is one of them...

Reply:
Ok just looked into it. According to the [BIP](https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki): 

"Inherited signaling: Transactions that don't explicitly signal replaceability are replaceable under this policy for as long as any one of their ancestors signals replaceability and remains unconfirmed"

*remains uncofirmed* is the key phrase here.
I will changes the implementation accordingly. Thanks for the sharp eye @iljakuklic